activerecord/lib/active_record/relation/predicate_builder/build_method.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class PredicateBuilder
+    class BuildMethod # :nodoc:
+      def initialize(klass)
+        @lookups = {}
+        klass&.aggregate_reflections&.keys&.each do |aggregate|
+          @lookups[aggregate] = :aggregate
+        end
+
+        # TODO: can I check `reflection.is_a?(ThroughReflection) instead?
+        klass&._reflections&.each do |reflection, reflection_obj|
+          @lookups[reflection] = :reflection
```

Comment:
We'll have the same `singluarize` problem with this approach, i.e. if the association is `comment` and the where key is `comments`, it won't be found.

Reply:
Yeah, I believe that is true. I think we can solve that in its own PR. It sounds like right now the answer is we can deprecate that behavior. I opened something for that here: https://github.com/rails/rails/pull/45163